File: src/game.py
"""
Game module for Battleship
Manages game state and rules
"""

from enum import Enum
from typing import Any, Dict, Optional

from src.player import Player


class GamePhase(Enum):
    """Game phases"""

    PLACEMENT = "placement"
    BATTLE = "battle"
    FINISHED = "finished"


class Game:
    """Manages a Battleship game"""

    def __init__(self, player1_name: str, player2_name: str):
        """
        Initialize a new game

        Args:
            player1_name: Name of player 1
            player2_name: Name of player 2
        """
        self.player1 = Player(player1_name)
        self.player2 = Player(player2_name)
        self.current_player = self.player1
        self.phase = GamePhase.PLACEMENT
        self.winner: Optional[Player] = None

    def get_opponent(self, player: Player) -> Player:
        """
        Get the opponent of a given player

        Args:
            player: The player whose opponent to get

        Returns:
            The opponent player
        """
        return self.player2 if player == self.player1 else self.player1

    def both_players_ready(self) -> bool:
        """
        Check if both players have placed all their ships

        Returns:
            True if both players are ready, False otherwise
        """
        return self.player1.all_ships_placed() and self.player2.all_ships_placed()

    def start_battle(self) -> None:
        """Start the battle phase"""
        if not self.both_players_ready():
            raise ValueError("Both players must place all ships before starting battle")
        self.phase = GamePhase.BATTLE

    def fire_shot(self, player: Player, row: int, col: int) -> Dict[str, Any]:
        """
        Fire a shot from one player to opponent

        Args:
            player: The player firing the shot
            row: Target row
            col: Target column

        Returns:
            Result of the shot (hit/miss/sunk info)

        Raises:
            ValueError: If not player's turn or wrong game phase
        """
        if self.phase != GamePhase.BATTLE:
            raise ValueError("Cannot fire shots during placement phase")

        if player != self.current_player:
            raise ValueError("Not your turn")

        opponent = self.get_opponent(player)
        result = opponent.receive_shot(row, col)

        # Check if game is over
        if opponent.all_ships_sunk():
            self.phase = GamePhase.FINISHED
            self.winner = player

        # Switch turns
        self.current_player = opponent

        return result

File: frontend/src/types/game.ts
export const CellState = {
  EMPTY: 'empty',
  SHIP: 'ship',
  HIT: 'hit',
  MISS: 'miss',
} as const;

export type CellState = typeof CellState[keyof typeof CellState];

export const ShipType = {
  CARRIER: 'carrier',
  BATTLESHIP: 'battleship',
  CRUISER: 'cruiser',
  SUBMARINE: 'submarine',
  DESTROYER: 'destroyer',
} as const;

export type ShipType = typeof ShipType[keyof typeof ShipType];

export const SHIP_LENGTHS: Record<ShipType, number> = {
  [ShipType.CARRIER]: 5,
  [ShipType.BATTLESHIP]: 4,
  [ShipType.CRUISER]: 3,
  [ShipType.SUBMARINE]: 3,
  [ShipType.DESTROYER]: 2,
};

export const Orientation = {
  HORIZONTAL: 'horizontal',
  VERTICAL: 'vertical',
} as const;

export type Orientation = typeof Orientation[keyof typeof Orientation];

export const GamePhase = {
  PLACEMENT: 'placement',
  BATTLE: 'battle',
  FINISHED: 'finished',
} as const;

export type GamePhase = typeof GamePhase[keyof typeof GamePhase];

export interface Ship {
  type: string;
  length: number;
  row: number;
  col: number;
  orientation: Orientation;
  hits: number;
  is_sunk: boolean;
}

export interface Player {
  name: string;
  grid: CellState[][];
  ships: Ship[];
  all_ships_placed: boolean;
  all_ships_sunk: boolean;
}

export interface GameState {
  game_id: string;
  mode: 'ai' | 'multiplayer';
  phase: GamePhase;
  current_turn: string;
  player1: Player;
  player2: Player;
  winner: string | null;
  created_at: string;
}

export interface ShotResult {
  result: 'hit' | 'miss' | 'sunk';
  ship_type?: ShipType;
  row: number;
  col: number;
}
